libvast/src/system/index.cpp

```diff
@@ -208,49 +208,149 @@ filesystem_actor& partition_factory::filesystem() {
+  auto it = queries.find(qid);
+  if (it == queries.end())
+    return caf::make_error(ec::unspecified, "cannot activate unknown query");
+  it->second.requested_partitions += num_partitions;
+  // Go over all currently inactive partitions and splice those relevant for
+  // `qid` back into the active queue.
+  auto new_inactive = std::vector<pending_queue::pq>{};
+  std::partition_copy(std::make_move_iterator(inactive_partitions.begin()),
+                      std::make_move_iterator(inactive_partitions.end()),
+                      std::back_inserter(partitions),
+                      std::back_inserter(new_inactive), [&](const auto& p) {
+                        return std::find(p.queries.begin(), p.queries.end(),
+                                         qid)
+                               != p.queries.end();
+                      });
+  inactive_partitions = std::move(new_inactive);
+  std::sort(partitions.begin(), partitions.end());
+  return caf::none;
+}
+
+[[nodiscard]] caf::error pending_queue::remove_query(const uuid& qid) {
+  // If we are currently working on this query we don't remove it from the
+  // list yet.
+  VAST_TRACE("index removes query {}", qid);
+  auto keep = false;
+  auto it = queries.find(qid);
+  if (it == queries.end())
+    return caf::make_error(ec::unspecified, "cannot remove unknown query");
+  if (it->second.completed_partitions < it->second.scheduled_partitions)
+    keep = true;
+  else {
+    queries.erase(it);
+  }
+  auto run = [&](auto& queue) {
+    auto it = queue.begin();
+    for (; it < queue.end(); ++it) {
+      auto queries_it = std::find(it->queries.begin(), it->queries.end(), qid);
+      if (queries_it == it->queries.end())
+        continue;
+      if (!keep) {
+        it->queries.erase(queries_it);
+        if (it->queries.empty())
+          it = queue.erase(it);
+      }
+    }
+  };
+  run(partitions);
+  run(inactive_partitions);
+  return caf::none;
+}
+
+// NOLINTNEXTLINE
+std::optional<pending_queue::pq> pending_queue::next() {
+  if (partitions.empty())
+    return std::nullopt;
+  auto result = partitions.back();
+  partitions.pop_back();
+  pq active = {result.partition, {}};
+  pq inactive = {result.partition, {}};
+  std::partition_copy(
+    std::make_move_iterator(result.queries.begin()),
+    std::make_move_iterator(result.queries.end()),
+    std::back_inserter(active.queries), std::back_inserter(inactive.queries),
+    [&](const auto& qid) {
+      const auto& qs = queries[qid];
+      return qs.requested_partitions > qs.scheduled_partitions;
+    });
+  if (!inactive.queries.empty())
+    inactive_partitions.push_back(std::move(inactive));
+  if (!active.queries.empty()) {
+    // TODO: Consider delaying the following loop until the next insert or
+    // activate.
+    for (const auto& qid : active.queries)
+      queries[qid].scheduled_partitions++;
+    return active;
   }
-  return std::nullopt;
+  return next();
```

Comment:
This is probably fine, but while thinking about it I found that clang has an `__attribute__((musttail))` that may be useful here.

Reply:
I just loopified the algorithm :shrug:.